lib/fuzzer/main.zig

```diff
@@ -0,0 +1,392 @@
+
+    if (remainder > 0) {
+        const i = pc_counters.len;
+        const elem = &seen_pcs[i];
+        var mask: usize = 0;
+        for (pc_counters_[i * @bitSizeOf(usize) ..][0..remainder], 0..) |byte, bit_index| {
+            mask |= @as(usize, @intFromBool(byte != 0)) << @intCast(bit_index);
+        }
+        _ = @atomicRmw(usize, elem, .Or, mask, .monotonic);
+    }
+}
+
+fn incrementUniqueRuns(seen_pcs: MemoryMappedList(u8)) void {
+    const header: *volatile SeenPcsHeader = @ptrCast(seen_pcs.items[0..@sizeOf(SeenPcsHeader)]);
+    _ = @atomicRmw(usize, &header.unique_runs, .Add, 1, .monotonic);
+}
+
+fn incrementNumberOfRuns(seen_pcs: MemoryMappedList(u8)) void {
+    const header: *volatile SeenPcsHeader = @ptrCast(seen_pcs.items[0..@sizeOf(SeenPcsHeader)]);
+    _ = @atomicRmw(usize, &header.n_runs, .Add, 1, .monotonic);
+}
+
+const InitialFeatureBufferCap = 64;
+
+pub const Fuzzer = struct {
+    gpa: Allocator,
+    rng: std.Random.DefaultPrng,
+    cache_dir: std.fs.Dir,
+
+    input_pool: InputPool,
+
+    mutate_scratch: ArrayListUnmanaged(u8) = .{},
+    mutation_seed: u64 = undefined,
+    mutation_len: usize = undefined,
+    current_input: ArrayListUnmanaged(u8) = .{},
+    current_input_checksum: u8 = undefined,
+
+    feature_buffer: []u32 = undefined,
+    all_features: ArrayListUnmanaged(u32) = .{},
+
+    // given to us by LLVM
+    pcs: []const usize,
+    pc_counters: []u8, // same length as pcs
+
+    n_runs: usize = 0,
+
+    /// Tracks which PCs have been seen across all runs that do not crash the fuzzer process.
+    /// Stored in a memory-mapped file so that it can be shared with other
+    /// processes and viewed while the fuzzer is running.
+    seen_pcs: MemoryMappedList(u8),
+
+    /// Identifies the file name that will be used to store coverage
+    /// information, available to other processes.
+    coverage_id: u64,
+
+    first_run: bool = true,
+
+    /// When we boot, we need to iterate over all corpus inputs and run them
+    /// once, populating initial feature set. When we are walking the corpus,
+    /// this variable stores current input index. After the walk is done, we
+    /// set it to null
+    corpus_walk: ?usize = null,
+
+    pub fn init(gpa: Allocator, cache_dir: std.fs.Dir, pc_counters: []u8, pcs: []usize) Fuzzer {
+        assert(pc_counters.len == pcs.len);
+
+        // Choose a file name for the coverage based on a hash of the PCs that
+        // will be stored within.
+        const pc_digest = hashPCs(pcs);
+        const coverage_id = pc_digest;
+        const hex_digest = std.fmt.hex(pc_digest);
+        const coverage_file_path = "v/" ++ hex_digest ++ "coverage";
+
+        const feature_buffer = check(@src(), gpa.alloc(u32, InitialFeatureBufferCap), .{});
+
+        const seen_pcs = initCoverageFile(cache_dir, coverage_file_path, pcs);
+
+        const input_pool = InputPool.init(cache_dir, pc_digest);
+
+        const seed: u64 = @bitCast(std.time.milliTimestamp());
```

Comment:
Never seed from the system time. `std.crypto.random` is the best place to get such a seed.

However, I intentionally seeded with 0 to make it deterministic, making it easier to evaluate changes to libfuzzer. What is your rationale for making it nondeterministic?

Reply:
> What is your rationale for making it nondeterministic

Sleep deprevation. Reverted :D